lib/surface.ex

```diff
@@ -248,6 +248,31 @@ defmodule Surface do
     []
   end
 
+  @doc false
+  def prop_to_opts(_module, _prop, nil) do
+    []
+  end
+
+  def prop_to_opts(module, prop, value) do
+    {type, _opts} = Surface.TypeHandler.attribute_type_and_opts(module, prop, nil)
+
+    {:safe, html} = Phoenix.HTML.raw(Surface.TypeHandler.attr_to_html!(type, prop, value))
+
+    # HACK:
+    # It seems like opts should be tuples of type and value e.g. {:class, "link primary"},
+    # But attr_to_html!\3 return the full html for the html attribute e.g. [" ", "class", "=", "\"", "link primary", "\""]
+    # How to we go from one form to the other?
+    html = Enum.at(html, -2)
+
+    # HACK:
+    # attribute_type_and_opts\3 gives us the internal surface type for a property e.g. :css_class
+    # How do we convert back from type to html attribute? E.g. :css_class -> :class
+    case type do
+      :css_class -> [{:class, html}]
+      _ -> [{type, html}]
+    end
+  end
```

Comment:
Yeah, you're right. `Surface.TypeHandler.attr_to_html!` is not exactly what we need. I think we could add a new function `Surface.TypeHandler.attr_to_opts!` that can convert the values directly to opts:

```elixir
  def attr_to_opts!(type, name, value) do
    case handler(type).value_to_html(name, value) do
      {:ok, val} when val in ["", nil, false] ->
        []

      {:ok, val} ->
        [{name, to_string(val)}]

      {:error, message} ->
        IOHelper.runtime_error(message)
    end
  end
```

and then the implementation of `Surface.prop_to_opts` would be:

```elixir
  @doc false
  def prop_to_opts(_prop, nil, _caller) do
    []
  end

  def prop_to_opts(prop, value, caller) do
    module = caller.module
    meta = %{caller: caller, line: caller.line, node_alias: module}
    {type, _opts} = Surface.TypeHandler.attribute_type_and_opts(module, prop, meta)
    Surface.TypeHandler.attr_to_opts!(type, prop, value)
  end
```

Regarding the inputs, we should also rely on the type handlers so we need to change the [get_non_nil_props](https://github.com/msaraiva/surface/blob/ea4ca563846e765bffed508e515dcec006e01baa/lib/surface/components/form/utils.ex#L11) function to this:

```elixir
  defmacro get_non_nil_props(assigns, props) do
    quote do
      unquote(__MODULE__).get_non_nil_props(unquote(assigns), unquote(props), __ENV__)
    end
  end

  def get_non_nil_props(assigns, props, caller) do
    module = caller.module
    meta = %{caller: caller, line: caller.line, node_alias: module}
    Enum.reduce(props, [], fn prop, acc ->
      {key, value} = prop_value(assigns, prop)
      {type, _opts} = Surface.TypeHandler.attribute_type_and_opts(module, key, meta)
      internal_value = Surface.TypeHandler.expr_to_value!(type, key, [value], [], module, inspect(value))
      Surface.TypeHandler.attr_to_opts!(type, key, internal_value) ++ acc
    end)
  end
```

After adding the above changes, everything should work but you will see a bunch of warnings regarding an undefined property `pattern` for component `TelephoneInput`. That's great! It shows how important is to rely on the type handlers. The warnings are correct since we forgot to declare the `pattern` property. Add this to `Surface.Components.Form.TelephoneInput`:

```elixir
  @doc "A regular expression to validate the entered value"
  property pattern, :string
```

and you should not see any warnings anymore. 

Let me know in case I've missed anything.

Reply:
Worked just as you described, thanks for the detailed guide!

I only made a small change from `prop_to_opts(prop, value, caller)` to `prop_to_opts(prop_value, prop_name, caller)`, to be more consistent with `event_to_opts/2`.